maggy/core/executors/tf_dist_executor.py

```diff
@@ -0,0 +1,370 @@
+                )
+
+            # Set retval to work with util.handle_return_value,
+            # if there is more than 1 metrics, retval will be a list and
+            # retval[0] will contain the final loss
+            retval = {"Metric": retval[0] if isinstance(retval, list) else retval}
+            retval = util.handle_return_val(retval, tb_logdir, "Metric", trial_log_file)
+            reporter.log("Finished distributed training.")
+            client.finalize_metric(retval, reporter)
+        except:  # noqa: E722
+            reporter.log(traceback.format_exc())
+            raise
+        finally:
+            reporter.close_logger()
+            client.stop()
+            client.close()
+
+    return wrapper_function
+
+
+def _register_with_servers(
+    client: Client, reporter: Reporter, partition_id: int
+) -> None:
+    """Registers own address with server and starts heartbeat protocol.
+
+    :param client: Client for communication with the server.
+    :param reporter: Reporter responsible for heartbeat.
+    :param partition_id: Executors partition ID from Sparks RDD.
+    """
+    client_addr = client.client_addr
+    port = _get_open_port()
+    host_port = client_addr[0] + ":" + str(port)
+    exec_spec = {
+        "partition_id": partition_id,
+        "task_attempt": 0,
+        "host_port": host_port,
+        "trial_id": None,
+    }
+    client.register(exec_spec)
+    client.start_heartbeat(reporter)
+
+
+def _get_open_port() -> str:
+    """Lets the OS choose a free socket and attempts to bind it.
+
+    :returns: The port name.
+    """
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.bind(("", 0))  # Bind to 0 lets OS choose a free socket.
+    port = sock.getsockname()[1]
+    sock.close()
+    return port
+
+
+def _setup_logging(reporter: Reporter, log_dir: str) -> Tuple[str, str]:
+    """Sets up logging directories and files.
+
+    :param reporter: Reporter responsible for logging.
+    :param log_dir: Log directory path on the file system.
+
+    :returns: Tuple containing the path of the tensorboard directory
+        and the trial log file.
+    """
+    reporter.set_trial_id(0)
+    tb_logdir = log_dir + "/" + "training_logs_" + str(reporter.partition_id)
+    trial_log_file = tb_logdir + "/output.log"
+    reporter.set_trial_id(0)
+    # If trial is repeated, delete trial directory, except log file
+    if EnvSing.get_instance().exists(tb_logdir):
+        util.clean_dir(tb_logdir, [trial_log_file])
+    else:
+        EnvSing.get_instance().mkdir(tb_logdir)
+    reporter.init_logger(trial_log_file)
+    return tb_logdir, trial_log_file
+
+
+def _setup_tf_config(tf_config: dict) -> None:
+    """Registers the TF_CONFIG environment variables on the worker.
+
+    :param tf_config: Dictionary containing the values of the variables.
+    """
+    os.environ["TF_CONFIG"] = json.dumps(tf_config)
+
+
+def _init_seed(random_seed: int = 0) -> None:
```

Comment:
where is this used and how do you make sure it's the same random seed on all executors?

Reply:
This is similar to what happens in maggy/core/executors/torch_dist_executor.py but since we don't need to initialize any cluter in tensorflow we just set the seed. Maybe a better name for the variable would be _set_seed().